### ISSUE TITLE
Support custom actions to be executed on a document type

### ISSUE THREAD

**Issue by OldSneerJaw** [Kashoo/synctos#25]
Allow a document definition to define fully custom action functions to be executed when the sync function is run. It should allow at least two different types:
- custom functions that are run after the document's type is identified, but before user authorization is verified (e.g. `customPreAction`)
- custom functions that are run after user authorization is verified and the document's contents pass validation, but before channels are assigned (e.g. `customPostAction`)

For example (not to be considered final):

```
myDocType: {
  channels: ...,
  typeFilter: ...,
  propertyValidators: ...,
  customPreAction: function(doc, oldDoc) {
    // Do something fancy before any validation is performed
  },
  customPostAction: function(doc, oldDoc) {
    // Do something fancy after all validation has passed
  }
}
```

**Comment by OldSneerJaw** [comment]
The implementation in pull request #62 implements custom action support for the following events:
* When a document's type is identified
* When authorization is granted to the user
* When document contents are validated
* When access is assigned to users/roles
* When channels are assigned to the document

**Comment by dkichler** [comment]
Addressed by #62 and #63
